File: Week1/2.Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, s1, s2):
        """
        :type s1: ListNode
        :type s2: ListNode
        :rtype: ListNode
        """
        
        ans = ListNode(0)
        temp = ans
        tempsum = 0
        while True:
            if s1 != None:
                tempsum += s1.val
                s1 = s1.next
            if s2 != None:
                tempsum += s2.val
                s2 = s2.next
            temp.val = tempsum % 10
            tempsum //= 10     #temp除以10的结果向负无穷大取整
            if s1 == None and s2 == None and tempsum == 0:
                break
            temp.next = ListNode(0)
            temp = temp.next
        return ans

File: Week1/2.Add Two Numbers.cpp
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head = new ListNode(-1);   //添加虚拟头结点，简化首尾的处理
        ListNode *cur = head;
        int carry = 0;  //表示进位
        while (l1 || l2) 
        {
            int n1 = l1 ? l1->val : 0;
            int n2 = l2 ? l2->val : 0;
            int sum = n1 + n2 + carry;
            carry = sum / 10;
            cur->next = new ListNode(sum % 10);
            cur = cur->next;
            if (l1) l1 = l1->next;
            if (l2) l2 = l2->next;
        }
        if (carry) cur->next = new ListNode(1); //如果最高位有进位，则需在最前面补1.
        return head->next;   //返回真正的头结点
    }
};

